state/spacesdiscovery.go

```diff
@@ -0,0 +1,162 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"github.com/juju/errors"
+	"github.com/juju/utils/set"
+	"gopkg.in/juju/names.v2"
+
+	"github.com/juju/juju/environs"
+	"github.com/juju/juju/instance"
+	"github.com/juju/juju/network"
+)
+
+func (st *State) getModelSubnets() (set.Strings, error) {
+	subnets, err := st.AllSubnets()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	modelSubnetIds := make(set.Strings)
+	for _, subnet := range subnets {
+		modelSubnetIds.Add(string(subnet.ProviderId()))
+	}
+	return modelSubnetIds, nil
+}
+
+// ReloadSpaces loads spaces and subnets from provider specified by environ into state.
+// Currently it's an append-only operation, no spaces/subnets are deleted.
+func (st *State) ReloadSpaces(environ environs.Environ) error {
+	netEnviron, ok := environs.SupportsNetworking(environ)
+	if !ok {
+		logger.Debugf("not a networking environ")
```

Comment:
So we call ReloadSpaces in a number of ways (always at every add space), so I don't know that this should be a higher warning level. But it would sort of be nice if the calls to Reload that were triggered by a user would indicate why we aren't updating any records.

Maybe we need to return an error, and have the built-in calls recognize it and discard it?
And then user-initiated calls can get a "space discovery not supported on this provider" ?